sr_robot_launch/src/sr_robot_launch/sr_ur_arm_unlock.py

```diff
@@ -0,0 +1,248 @@
+#!/usr/bin/env python3
+# Copyright 2020, 2021 Shadow Robot Company Ltd.
+#
+# This program is free software: you can redistribute it and/or modify it
+# under the terms of the GNU General Public License as published by the Free
+# Software Foundation version 2 of the License.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program. If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import absolute_import
+import rospy
+from std_msgs.msg import Bool
+from ur_dashboard_msgs.srv import GetSafetyMode, GetProgramState, GetRobotMode, Load, IsProgramRunning
+from ur_dashboard_msgs.msg import SafetyMode, ProgramState, RobotMode
+from std_srvs.srv import Trigger
+
+
+class RobotSafetyMonitor(object):
+    def __init__(self, name):
+        self.name = name
+        self.substring = None
+        self.subscriber = None
+        self.estop_pressed = False
+        self.mode = None
+
+    def setup_subscriber(self):
+        self.subscriber = rospy.Subscriber(self.substring, SafetyMode,
+                                           self.safety_mode_callback)
+
+    def press_estop(self):
+        self.estop_pressed = True
+
+    def release_estop(self):
+        self.estop_pressed = False
+
+    def safety_mode_callback(self, message):
+        self.mode = message.mode
+        if message.mode == SafetyMode.ROBOT_EMERGENCY_STOP:
+            self.press_estop()
+
+
+class SrUrUnlock(object):
+    def __init__(self):
+        self._external_control_program_name = rospy.get_param("~urcap_program_name", "external_ctrl.urp")
+        self._arms = []
+        if rospy.has_param('ra_sr_ur_robot_hw'):
+            self._arms.append('ra')
+        if rospy.has_param('la_sr_ur_robot_hw'):
+            self._arms.append('la')
+        if 0 == len(self._arms):
+            rospy.logerr("No arms detected, shutting down %s", rospy.get_name())
+            rospy.signal_shutdown("No arms detected")
+        rospy.Subscriber("/sr_arm/release_or_brake", Bool, self.release_or_brake_arm_cb)
+        self._robot_state_monitors = {}
+        self.setup_robot_mode_subscribers()
+
+    def setup_robot_mode_subscribers(self):
+        for arm in self._arms:
+            robot_safety_monitor = RobotSafetyMonitor(arm)
+            robot_safety_monitor.substring = '/' + arm + '_sr_ur_robot_hw/safety_mode'
```

Comment:
What is the difference between the 'name' and 'substring'? I can't see where you use 'name'. I think within the constructor you can define your' substring instead of assigning it externaly

Reply:
yeah that's a better way of doing it, thanks

Reply:
71d2bc6d3854b4816d8790a2a364fe157085ec51

Reply:
Thanks